File: src/test/serviceEngine.test.ts
/**
 * Slice-B Phase-7 — serviceEngine ranking tests.
 */
import { describe, it, expect } from "vitest";
import { rankServices } from "@/core/serviceEngine";

describe("serviceEngine.rankServices", () => {
  it("boosts visa+insurance for foreign trips in <30 days", () => {
    const r = rankServices({
      activeCountryIso2: "US",
      nextDestinationIso2: "GB",
      nextTrip: null,
      budget: null,
      daysToNextTrip: 14,
    });
    const visaIdx = r.findIndex((x) => x.tab === "visa");
    const localIdx = r.findIndex((x) => x.tab === "local");
    expect(visaIdx).toBeLessThan(localIdx);
  });

  it("boosts rides+food when trip imminent", () => {
    const r = rankServices({
      activeCountryIso2: "US",
      nextDestinationIso2: "FR",
      nextTrip: null,
      budget: null,
      daysToNextTrip: 1,
    });
    const top = r.slice(0, 3).map((x) => x.tab);
    expect(top).toContain("rides");
  });

  it("favours local when no upcoming trip", () => {
    const r = rankServices({
      activeCountryIso2: "JP",
      nextDestinationIso2: null,
      nextTrip: null,
      budget: null,
      daysToNextTrip: Number.POSITIVE_INFINITY,
    });
    const local = r.find((x) => x.tab === "local");
    expect(local).toBeDefined();
    expect(local!.score).toBeGreaterThan(1);
  });

  it("returns deterministic ordering for identical inputs", () => {
    const i = {
      activeCountryIso2: "US",
      nextDestinationIso2: "GB",
      nextTrip: null,
      budget: null,
      daysToNextTrip: 5,
    };
    const a = rankServices(i);
    const b = rankServices(i);
    expect(a.map((x) => x.tab)).toEqual(b.map((x) => x.tab));
  });
});

File: flutter_app/test/service_engine_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:globeid/domain/service_engine.dart';

void main() {
  group('service engine', () {
    test('ranks visa and insurance for upcoming foreign travel', () {
      final rankings = rankServices(const ServiceInput(
        activeCountryIso2: 'US',
        nextDestinationIso2: 'JP',
        daysToNextTrip: 14,
        overBudgetCategoryCount: 0,
      ));

      expect(rankings.first.tab, ServiceTab.visa);
      expect(
        rankings.take(4).map((r) => r.tab),
        containsAll(<ServiceTab>[
          ServiceTab.visa,
          ServiceTab.insurance,
          ServiceTab.esim,
          ServiceTab.exchange,
        ]),
      );
    });

    test('boosts ride urgency for imminent trips', () {
      final rankings = rankServices(const ServiceInput(
        activeCountryIso2: 'IN',
        nextDestinationIso2: 'AE',
        daysToNextTrip: 1,
        overBudgetCategoryCount: 0,
      ));

      expect(rankings.first.tab, ServiceTab.rides);
      expect(rankings.first.reason, contains('book the ride'));
    });

    test('surfaces local services when already abroad without a near trip', () {
      final rankings = rankServices(const ServiceInput(
        activeCountryIso2: 'SG',
        nextDestinationIso2: null,
        daysToNextTrip: 9999,
        overBudgetCategoryCount: 0,
      ));

      expect(rankings.first.tab, ServiceTab.local);
      expect(rankings[1].tab, ServiceTab.food);
    });

    test('nudges exchange when budget categories are over', () {
      final rankings = rankServices(const ServiceInput(
        activeCountryIso2: 'US',
        nextDestinationIso2: 'US',
        daysToNextTrip: 40,
        overBudgetCategoryCount: 2,
      ));

      final exchange = rankings.firstWhere((r) => r.tab == ServiceTab.exchange);
      final hotels = rankings.firstWhere((r) => r.tab == ServiceTab.hotels);
      expect(exchange.score, greaterThan(hotels.score));
    });

    test('maps airport country and destination currency', () {
      expect(airportCountryIso2('hnd'), 'JP');
      expect(currencyForCountryIso2('jp'), 'JPY');
      expect(airportCountryIso2('unknown'), isNull);
    });
  });
}
